### PR TITLE
Add llama3.1 8b mxfp4 support

### PR DESCRIPTION
**Pull request by vidushi8**
Add support for llama3.1 8b mxfp4 through TE backend.

Training loss curve comparing fp8, mxfp8 and mxfp4 following closely and converging. Used c4 dataset

image

Training cmd:
`NVTE_USE_CAST_TRANSPOSE=0 NVTE_MXFP4_USE_HADAMARD=1 bash runner/primus-cli direct   -- train pretrain   --config examples/megatron/configs/MI355X/llama3.1_8B-MXFP4-pretrain.yaml --train_iters 50 `

### CODE REVIEW

File: examples/megatron/configs/MI355X/llama3.1_8B-MXFP8-pretrain.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      use_distributed_optimizer: true # this is needed for fsdp2
+
+      # Cross entropy flags
+      cross_entropy_fusion_impl: "te"
+      cross_entropy_loss_fusion: true
+
+      # --- Mixed Precision Config ---
+      fp8: e4m3
+      fp8_recipe: mxfp8  
+
```
[the remark this answers is not in the document]
Comment (vidushi8) [reply examples/megatron/configs/MI355X/llama3.1_8B-MXFP8-pretrain.yaml:105]: mxfp4 yaml added as well `examples/megatron/configs/MI355X/llama3.1_8B-MXFP4-pretrain.yaml`

File: examples/megatron/configs/MI355X/llama3.1_8B-MXFP8-pretrain.yaml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      gradient_accumulation_fusion: true
+
+      # --- Data Config ---
+      mock_data: true
+      train_data_path: null
+      valid_data_path: null
+      test_data_path: null
+      seq_length: 8192
+      data_cache_path: null
+      mmap_bin_files: true
```
[the remark this answers is not in the document]
Comment (vidushi8) [reply examples/megatron/configs/MI355X/llama3.1_8B-MXFP8-pretrain.yaml:68]: fixed

File: examples/megatron/configs/MI355X/llama3.1_8B-MXFP8-pretrain.yaml [hunk 3]
```diff
@@ -0,0 +1,118 @@
+work_group: ${TEAM:amd}
+user_name: ${USER:root}
+exp_name: ${EXP_NAME:llama3.1_8B-pretrain-mxfp8}
+workspace: ./output
```
[the remark this answers is not in the document]
Comment (vidushi8) [reply examples/megatron/configs/MI355X/llama3.1_8B-MXFP8-pretrain.yaml]: fixed

File: primus/backends/megatron/patches/turbo/fp4_patches.py [hunk 4]
```diff
@@ -20,12 +20,17 @@ def _is_fp4_can_patch(ctx: PatchContext) -> bool:
     return fp4 and is_primus_turbo_can_patch(ctx)
 
 
+def _is_fp4_enabled(ctx: PatchContext) -> bool:
+    args = get_args(ctx)
+    return bool(getattr(args, "fp4", False))
```
[the remark this answers is not in the document]
Comment (vidushi8) [reply primus/backends/megatron/patches/turbo/fp4_patches.py:25]: keeping for legacy purposes. `_is_fp4_can_patch()` check for turbo. fp4 can also use TE only path, does not have to always use turbo.

File: primus/backends/megatron/core/fp4_utils.py [hunk 5]
```diff
@@ -130,13 +150,31 @@ def get_fp4_context(config: TransformerConfig, layer_no: int = -1, is_init: bool
...
+                        fp4_recipe=fp4_recipe,
+                        fp4_group=fp4_group,
+                        enabled_turbo=True if fp4_quant_config is not None else False,
+                        turbo_quant_config=fp4_quant_config,
+                    )
+                else:
+                    # TE currently uses fp8_autocast for fp8 and fp4 quantization.
+                    fp4_context = transformer_engine.pytorch.fp8_autocast(
+                        enabled=True, fp8_recipe=fp4_recipe, fp8_group=fp4_group
+                    )
```
[the remark this answers is not in the document]
Comment (vidushi8) [reply primus/backends/megatron/core/fp4_utils.py:184]: Outdate comment. Already fixed

File: primus/backends/megatron/core/fp4_utils.py [hunk 6]
```diff
@@ -151,25 +189,24 @@ def get_fp4_context(config: TransformerConfig, layer_no: int = -1, is_init: bool
...
                 )
+            fp4_recipe = transformer_engine.common.recipe.NVFP4BlockScaling()
+        elif config.fp4_recipe == Fp4Recipe.mxfp4:
+            if not is_te_min_version("2.8.0"):
+                raise ValueError(
+                    "MXFP4BlockScaling requires TransformerEngine >= 2.8.0."
+            )
+            import os       
+            fp4_recipe = transformer_engine.common.recipe.MXFP4BlockScaling()
+            fp4_recipe.use_hadamard = os.environ.get("NVTE_MXFP4_USE_HADAMARD", "0") == "1"
```
[the remark this answers is not in the document]
Comment (vidushi8) [reply primus/backends/megatron/core/fp4_utils.py]: outdate. Already fixed

**Review by wenxie-amd — APPROVED** [review]
